infer/lib/audio.py

```diff
@@ -39,19 +40,102 @@ def load_audio(file: str, sr: int) -> np.ndarray:
         raise FileNotFoundError(f"File not found: {file}")
 
     try:
-        # https://github.com/openai/whisper/blob/main/whisper/audio.py#L26
-        # This launches a subprocess to decode audio while down-mixing and resampling as necessary.
-        # Requires the ffmpeg CLI and `ffmpeg-python` package to be installed.
-        file = str(clean_path(file))  # 防止小白拷路径头尾带了空格和"和回车
-        out, _ = (
-            ffmpeg.input(file, threads=0)
-            .output("-", format="f32le", acodec="pcm_f32le", ac=1, ar=sr)
-            .run(cmd=["ffmpeg", "-nostdin"], capture_stdout=True, capture_stderr=True)
-        )
+        container = av.open(file)
+        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)
+
+        # AV stores duration in nanoseconds
+        decoded_audio = (
+            ((container.duration * sr / container.bit_rate) // 1_000_000) + 1
+        ) * []
+
+        for frame in container.decode(audio=0):
+            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
+            resampled = resampler.resample(frame)
+            decoded_audio.append(np.array(resampled))
+
+        audio = np.concatenate(decoded_audio)
     except Exception as e:
         raise RuntimeError(f"Failed to load audio: {e}")
 
-    return np.frombuffer(out, np.float32).flatten()
+    return np.frombuffer(audio, dtype=np.float32).flatten()

```

Comment:
If the dtype of decoded_audio is already float32, now no need to call frombuffer and flatten, just return decoded_audio.

Reply:
so, something like this?
```py
return np.frombuffer(audio, dtype=np.float32).flatten() if not isinstance(audio, np.float32) else audio
```
or this?
```py
return np.frombuffer(audio, dtype=np.float32).flatten() if not isinstance(audio, np.ndarray) else audio
```

Reply:
No, you just alloc a numpy array with dtype=np.float32 as decoded_audio, so no need to check anymore.

Reply:
> No, you just alloc a numpy array with dtype=np.float32 as decoded_audio, so no need to check anymore.

So... basically declare decoded_audio like so:
```py
decoded_audio = np.array([], dtype=np.float32)
```
?

Reply:
You can use np.zeros to allocate an array. Try to google or ask GPT are some good choices. 👍

Reply:
> You can use np.zeros to allocate an array. Try to google or ask GPT are some good choices. 👍

I'll come back to this tomorrow, I'm EXTREMELY sleep deprived at the moment :sweat_smile: 

```py
def load_audio(file: str, sr: int) -> np.ndarray:
    if not hasattr(load_audio, "offset"):
        load_audio.offset = 0 # Cache offset between function calls
    
    if not Path(file).exists():
        raise FileNotFoundError(f"File not found: {file}")

    try:
        container = av.open(file)
        resampler = AudioResampler(format="fltp", layout="mono", rate=sr)

        num_samples = int(((container.duration * sr / container.bit_rate) // 1_000_000) + 1)
        print(num_samples)
        decoded_audio = np.zeros(num_samples, dtype=np.float32)

        for frame in container.decode(audio=0):
            frame.pts = None  # Clear presentation timestamp to avoid resampling issues
            resampled = np.array(resampler.resample(frame))

            frame_length = resampled.shape[0]
            decoded_audio[load_audio.offset:load_audio.offset+frame_length] = resampled
            load_audio.offset += frame_length

        container.close()
    except Exception as e:
        raise RuntimeError(f"Failed to load audio: {e}")

    return decoded_audio[:load_audio.offset].flatten()
```

```bash
  File "/kirby/Coding/Retrieval-based-Voice-Conversion-WebUI/infer/lib/audio.py", line 63, in load_audio
    raise RuntimeError(f"Failed to load audio: {e}")
RuntimeError: Failed to load audio: float() argument must be a string or a real number, not 'av.audio.frame.AudioFrame'
```

Reply:
> I'll come back to this tomorrow, I'm EXTREMELY sleep deprived at the moment

You can do these things anytime you like. No strict deadline in this project. Of course, you don't need to reply to my msgs immediately.

Reply:
Assuming this was resolved with 953b2d6